src/config.rs

```diff
@@ -2,16 +2,118 @@ pub mod settings;
 pub mod storage;
 
 use crate::config::settings::{Setting, SettingInfo};
-use crate::types::Result;
+use crate::config::storage::memory::MemoryStorageAdapter;
+use lazy_static::lazy_static;
+use log::warn;
 use serde_derive::Deserialize;
 use serde_json::Value;
 use std::collections::HashMap;
 use std::mem;
 use std::str::FromStr;
+use std::sync::RwLock;
 use wildmatch::WildMatch;
 
+/// Settings are stored in a json file, but this is included in the binary for mostly easy editting.
 const SETTINGS_JSON: &str = include_str!("./config/settings.json");
 
+/// StoreAdapter is the interface for storing and retrieving settings
+/// This can be used to storage settings in a database, json file, etc
+/// Note that we need to implement Send so we can send the storage adapter
+/// to other threads.
+pub trait StorageAdapter: Send + Sync {
```

Comment:
Send + Sync are important here, otherwise we cannot share them with other threads. We must make sure the adapters are thread-aware.